java/core/src/main/java/com/google/protobuf/DescriptorMessageInfoFactory.java

```diff
@@ -593,21 +616,88 @@ static String getFieldName(FieldDescriptor fd) {
     String name = (fd.getType() == FieldDescriptor.Type.GROUP)
                   ? fd.getMessageType().getName()
                   : fd.getName();
-    String suffix = specialFieldNames.contains(name) ? "__" : "_";
-    return snakeCaseToCamelCase(name) + suffix;
+
+    // convert to UpperCamelCase for comparison to the specialFieldNames
+    // (which are in UpperCamelCase)
+    String upperCamelCaseName = snakeCaseToUpperCamelCase(name);
+
+    String suffix = specialFieldNames.contains(upperCamelCaseName)
+                  // For field names that match the specialFieldNames,
+                  // append "__" to prevent field accessor method names from
```

Comment:
Why both __ and _? Is there a conflict if both cases uses a single underscore?

Reply:
The single and double underscores referenced here are simply to match the java code that is currently generated by the protoc java compiler.  Notice the previous code used the single and double underscore.  In other words, my change does not change the existing semantics for single and double underscores.

The protoc java compiler currently (prior to my change) uses a single or double underscore depending on the proto field name.

The protoc java compiler always appends at least one underscore to java field names.  This prevents those field names from clashing with java keywords.  For example, if the proto field name was `int`, then the protoc java compiler would generate a java field name of 'int_'.

The protoc java compiler does not include the last trailing underscore in accessor method names for these fields.  For example, the getter method for the proto field named `int` would be `getInt()`.

In addition, the protoc java compiler adds _another_ underscore (for a total of two underscores) to the java field name for names that would result in accessor method names that clash with other method names.  For example, if the proto field name was `class`, the java field name would be `class__` (double underscore).  And again, the last trailing underscore is not included in the accessor method names.  So the getter method for the proto field named `class` would be `getClass_()`

This behavior existed prior to my change.   My change simply adds _more_ detection of field names that result in accessor method name clashes.  Previously, only `cached_size`, `serialized_size`, and `class` were detected.  Unfortunately, those are only a subset of the field names that would actually cause assessor method name conflicts.  My change adds the complete set of field names that cause accessor method name conflicts.

Again, I'm not changing the semantics of single or double underscores.  I am also not going to refactor the protoc compiler to change its current semantics for single and double underscores.  I'm simply fixing a bug in the protoc compiler where it was generating non-compilable java code due to missing detection of field names that cause conflicts.